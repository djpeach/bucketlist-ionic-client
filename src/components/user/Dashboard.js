import React from 'react';
import {
  IonPage,
  IonContent,
  IonButton,
  IonModal,
  IonTitle,
  IonCard,
  IonCardHeader,
  IonItem,
  IonList,
  IonLabel,
  IonSelect,
  IonSelectOption,
  IonInput,
} from "@ionic/react";
import authedComponent from "../common/AuthedComponent";
import NewDropsPreview from './NewDropsPreview'
import ListsPreview from './ListsPreview'
import {useState} from 'react'
import { useQuery, useMutation } from '@apollo/react-hooks';
import gql from '../../graphql'
import firebase from 'firebase'

function ListOfLists({onSelected}) {
  const {loading, error, data} = useQuery(gql.getListsByUser, {variables: {id: firebase.auth().currentUser.uid}})

  if (loading) {
    return (
      <IonList>
        <IonItem>
          <IonLabel>
            Loading Lists ...
          </IonLabel>
        </IonItem>
      </IonList>
    )
  }

  if (error) {
    return (
      <IonList>
        <IonItem>
          <IonLabel>
           Error: {error.message}
          </IonLabel>
        </IonItem>
      </IonList>
    )
  }

  return (
    <IonList>
      <IonItem>
        {data.getListsByUser.length > 0 ? (
            <>
              <IonLabel>Your Buckets: </IonLabel>
              <IonSelect placeholder="Select a Bucket" okText="Select" cancelText="Cancel" onIonChange={(e) => onSelected(e.target.value)}>
                {data.getListsByUser.map((list, index) => {
                  return (
                      <IonSelectOption value={list.id} key={list.id}>
                        {list.title}
                      </IonSelectOption>
                  )
                })}
              </IonSelect>
            </>
        ) : (
          <IonLabel>
          No Buckets yet, go back and make one!
          </IonLabel>
        )}
      </IonItem>
    </IonList>
  )
}

function BucketSelectModal({acceptingItem, drop, setAcceptingItem}) {
  const [selectedBucket, changeSelectedBucket] = useState('')
  const [assignItemToList] = useMutation(gql.assignItemToList, {
    onCompleted() {
      setAcceptingItem(false)
    }
  })

  return (
    <IonModal isOpen={acceptingItem}>
      <IonContent>
        <h1 className="bl-list-title">Add New Drop to a Bucket</h1>
        <IonCard className="mt-5">
          <IonCardHeader>New Drop</IonCardHeader>
          <IonItem>
            {drop !== {} ? (
              <div>
                <p>From: {drop.from && drop.from.firstName} {drop.from && drop.from.lastName}</p>
                <h5>{drop.message}</h5>
              </div>
            ) : (
              <p>No Drops</p>
            )}
          </IonItem>
        </IonCard>
        <IonCard>
          <IonCardHeader>Select a Bucket</IonCardHeader>
          <ListOfLists onSelected={changeSelectedBucket}/>
        </IonCard>
        <IonButton className="bl-list-back-btn" onClick={() => {
          assignItemToList({variables: {
            id: drop.id,
            listId: selectedBucket
          }})
        }}>
          Add to Bucket
        </IonButton>
        <IonButton className="fix-to-bottom" color="danger" onClick={() => setAcceptingItem(false)}>Cancel</IonButton>
      </IonContent>
    </IonModal>
  )
}

function Dashboard() {
  const [acceptingItem, setAcceptingItem] = useState(false)
  const [drop, setDrop] = useState({})
  const [creatingBucket, setCreatingBucket] = useState(false)
  const [bucketName, setBucketName] = useState('')
  const [createList] = useMutation(gql.createList, {
    onCompleted() {
      setCreatingBucket(false)
      setBucketName('')
    }
  })

  return (
    <IonPage className="bl-page">
      <IonContent fullscreen>
        <BucketSelectModal acceptingItem={acceptingItem} drop={drop} setAcceptingItem={setAcceptingItem}/>
        <NewDropsPreview setAcceptingItem={setAcceptingItem} setDrop={setDrop}/>
        <ListsPreview/>
        {creatingBucket && <IonInput placeholder="Bucket Name" value={ bucketName }
               onInput={ e => setBucketName(e.target.value) } />}
        <IonButton color="success" strong type="button"
                  className="ion-float-right ion-margin-end ion-margin-bottom bl-new-list-btn" onClick={() => {
                    if (creatingBucket) {
                      if (bucketName !== '') {
                        createList({variables: {
                          title: bucketName,
                          userId: firebase.auth().currentUser.uid
                        }})
                      }
                    } else {
                      setCreatingBucket(true)
                    }
                  }}>
          {creatingBucket ? 'Add Bucket' : '+ New Bucket'}
        </IonButton>
      </IonContent>
    </IonPage>
  )
};

export default authedComponent(Dashboard);