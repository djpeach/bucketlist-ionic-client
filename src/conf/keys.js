export default {
  firebaseConfig: {
    apiKey: 'AIzaSyD2zEhOKEYPM9dE3k4D1LQHw_hhBGgWjDk',
    authDomain: 'bucketlist-96454.firebaseapp.com',
    databaseURL: 'https://bucketlist-96454.firebaseio.com',
    projectId: 'bucketlist-96454',
    storageBucket: 'bucketlist-96454.appspot.com',
    messagingSenderId: '353678555750',
    appId: '1:353678555750:web:fdb8db821d026b8c86cba7',
  },
}
