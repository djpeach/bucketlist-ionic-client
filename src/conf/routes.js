export default {
  index: '/',
  home: '/home',
  more: '/more',
  lists: {
    create: '/lists/new',
    list: '/lists',
    detail: '/lists/:id',
    delete: '/lists/delete',
  },
  suggestions: {
    create: '/suggestions/new',
    list: '/suggestions',
    detail: '/suggestions/:id',
    delete: '/suggestions/delete',
  },
  auth: {
    login: '/auth/login',
    register: '/auth/register',
  },
  addFriend: '/friends/new',
}

export const routeWithParams = (route, ...params) => {
  const re = /(?<=\/):\w+/g
  let i = 0
  let match
  while ((match = re.exec(route)) != null) {
    route = route.replace(match[0], params[i])
    i++
  }
  return route
}
