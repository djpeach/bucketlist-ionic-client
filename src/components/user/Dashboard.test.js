import React from 'react'
import { shallow } from 'enzyme'
import firebase from 'firebase'
import keys from '../../conf/keys'
import routes from '../../conf/routes'


firebase.initializeApp(keys.firebaseConfig)

import Dashboard from './Dashboard'

let wrapper
let state

describe('Dashboard unit tests', () => {
  beforeEach(() => {
    wrapper = shallow(<Dashboard />)
  })
  it('should render', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('redirects to login if not logged in', () => {
    expect(wrapper.find('Redirect').prop('to')).toBe("/auth/login")
  })
})
