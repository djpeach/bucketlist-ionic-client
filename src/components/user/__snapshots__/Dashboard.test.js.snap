// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard unit tests should render 1`] = `
<Redirect
  to="/auth/login"
/>
`;
